File: 백준/Silver/15665. N과 M （11）/N과 M （11）.js
/*
이미 중복인 숫자는 넣지 않는다. 
이전 숫자와 중복인 숫자는 괜찮다.
*/

const filePath = process.platform === 'linux' ? '/dev/stdin' : 'e.txt';
const input = require('fs').readFileSync(filePath).toString().trim().split('\n');

const [count, maxLength] = input[0].split(' ').map(Number);
const numbers = input[1]
  .split(' ')
  .map(Number)
  .sort((a, b) => a - b);

const temp = [];
const resultList = [];
function bt() {
  if (temp.length === maxLength) {
    resultList.push(temp.join(' '));
    return;
  }

  let value = null;
  for (let i = 0; i < count; i++) {
    if (value === numbers[i]) continue;
    value = numbers[i];

    temp.push(numbers[i]);
    bt();
    temp.pop();
  }
}

bt();

console.log(resultList.join('\n'));

File: 백준/Silver/15665. N과 M （11）/N과 M （11）.py
n, m = map(int, input().split())
lst = list(map(int, input().split()))

lst.sort()

temp = []
def perm():
    if len(temp) == m:
        print(' '.join(map(str, temp)))
        return
    
    prev = None
    for i in range(n):
        cur = lst[i]
        
        if cur == prev:
            continue
        
        prev = cur
        
        temp.append(cur)
        perm()
        temp.pop()
    
perm()